coderd/notifications/dispatch/smtp.go

```diff
@@ -431,6 +432,15 @@ func (s *SMTPHandler) loadCertificate() (*tls.Certificate, error) {
 func (s *SMTPHandler) auth(ctx context.Context, mechs string) (sasl.Client, error) {
 	username := s.cfg.Auth.Username.String()
 
+	// All auth mechanisms require username, so if one is not defined then don't return an auth client.
+	if username == "" {
+		s.noAuthWarnOnce.Do(func() {
+			s.log.Warn(ctx, "skipping auth; no username configured", slog.F("support_mechanisms", mechs))
+		})
+		// nolint:nilnil // This is a valid response.
+		return nil, nil
```

Comment:
Is the new return signature handled by the caller(s)? I see that nilnil is new (unless sasl lib also returns it).

Reply:
Yup.

```go
		// Check for authentication capabilities.
		if ok, avail := c.Extension("AUTH"); ok { <--- enters if auth is supported
			// Ensure the auth mechanisms available are ones we can use.
			auth, err := s.auth(ctx, avail)
			if err != nil {
				return true, xerrors.Errorf("determine auth mechanism: %w", err)
			}

			// If so, use the auth mechanism to authenticate.
			if auth != nil { <--- bails out because auth is nil
				if err := c.Auth(auth); err != nil {
					return true, xerrors.Errorf("%T auth: %w", auth, err)
				}
			}
		} else if !s.cfg.Auth.Empty() {
			return false, xerrors.New("no authentication mechanisms supported by server")
		}

		<--- ... continues to process message without auth ...
```

Reply:
Admittedly the comments could be better about this case, lemme add that.

Reply:
See f7fcaf92a @mafredri, thanks for the prompt.